internal_ws/ykcompile/src/arch/x86_64/mod.rs

```diff
@@ -123,45 +123,121 @@ macro_rules! binop_add_sub {
+    //   - Both MUL and DIV's first (implicit) operand is twice as large as the second (explicit)
+    //     operand. For all but r/m8 variants, the operand extends into RDX.
+    //   - Similarly, the result of both MUL and DIV is twice as large as the second operand.
+    //
+    // SIR only deals with same-sized operands and doesn't require us to capture the remainder for
+    // division, so we don't use this extended operand/result space. This makes MUL and DIV very
+    // similar from our point of view, hence this macro is used to generate both implementations.
+    //
+    // We are however, required to zero the unused extended operand space to get the correct
+    // result, and we must be careful to save and restore anything that would be clobbered.
+    // Specifically:
+    //
+    //   - RAX is always clobbered, so we must always save and restore it.
+    //   - For all but the r/m8 variations, RDX is clobbered. So if the size of our SIR operands is
+    //     >1 byte, then we must also save and restore RDX.
+    //   - If our register allocator has put the second SIR operand in a register that will be
+    //     clobbered, then we cannot use this register as the r/m operand. In this case we borrow
+    //     RCX temporarily.
     ($name: ident, $op:expr) => {
         fn $name(&mut self, opnd1_reg: u8, opnd2: &IRPlace) {
-            // mul and div overwrite RAX, RDX, so save them first.
+            let size = SIR.ty(&opnd2.ty()).size();
+            // For all but r/m8 variants, RDX is clobbered.
+            if size > 1 {
+                dynasm!(self.asm
+                    ; push rdx
+                );
+            }
+            // RAX is an always clobbered.
             dynasm!(self.asm
                 ; push rax
-                ; push rdx
-                ; xor rdx, rdx
             );
+            // Set up first operand.
             dynasm!(self.asm
                 ; mov rax, Rq(opnd1_reg)
             );
-            let size = SIR.ty(&opnd2.ty()).size();
+            // Set up second operand.
             let src_loc = self.iplace_to_location(opnd2);
             match src_loc {
-                Location::Reg(r) => match size {
-                    1 => {
+                Location::Reg(src_r) => {
+                    // Handle cases where the second operand was stored in an already clobbered
+                    // register. In these cases we load the value back from the stack.
+                    let (r, borrow_rcx) = if src_r == RAX.code() {
+                        let rax_off = i32::try_from(QWORD_REG_SIZE).unwrap();
                         dynasm!(self.asm
-                            ; $op Rb(r)
+                            ; push rcx
+                            ; mov rcx, [rsp + rax_off]
                         );
-                    }
-                    2 => {
+                        (RCX.code(), true)
+                    } else if size > 1 && src_r == RDX.code() {
```

Comment:
Why do we need to do this? `rdx` hasn't been clobbered yet. Do we need to do this because we are `xor`ing `rdx` further below. In any case can we just do `mov rcx, rdx` here, as it looks like `rdx` is still intact at this point.

Reply:
You're right. Fixed in Fixed in 0214bfc9.